Add WorkerThreadPool documentation

MewPurPur opened the pull request:
Supersedes #65268. I made this PR without knowing about his work.

MewPurPur committed:
Add WorkerThreadPool documentation

mhilbrunner commented:
cc @RandomShaper @raulsntos Any preference on which of these two to merge? Does this one look good? :)

raulsntos commented on doc/classes/WorkerThreadPool.xml:
```diff
@@ -1,10 +1,50 @@
+		    # Other code that depends on the enemy AI already being processed.
+		[/gdscript]
+		[csharp]
+		private Node[] enemies = [] // An array of enemies that must be processed every frame.
```
```suggestion
		private Node[] _enemies = {}; // An array of enemies that must be processed every frame.
```

For empty arrays it's better to use `Array.Empty<Node>()` instead of `{}`. However, I assume this array is not meant to be empty. Using `Godot.Collections.Array` of `System.List` may be better, but I don't know where the enemies are suppose to come from.

raulsntos commented on doc/classes/WorkerThreadPool.xml:
```diff
@@ -1,10 +1,50 @@
+		[csharp]
+		private Node[] enemies = [] // An array of enemies that must be processed every frame.
+
+		private void processEnemyAI(int enemyIndex)
```
.NET naming guidelines recommend PascalCase for methods.

```suggestion
		private void ProcessEnemyAI(int enemyIndex)
```

raulsntos commented on doc/classes/WorkerThreadPool.xml:
```diff
@@ -1,10 +1,50 @@
+
+		public override void _Process(double delta)
+		{
+		    int taskID = WorkerThreadPool.addGroupTask(processEnemyAI, enemies.size())
+		    // Other code...
+		    WorkerThreadPool.waitForTaskCompletion(taskID)
+		    // Other code that depends on the enemy AI already being processed.
```
The Godot C# API follows the .NET naming conventions so method names are PascalCase.

If using an array (`Node[]`) the equivalent to GDScript's `.size()` would be `.Length`. If using `Godot.Collections.Array` or `System.List`, the equivalent would be `.Count`.

Also, the method group can't be implicitly converted to a `Callable`, so we need to use the static `From` method.

```suggestion
		    long taskID = WorkerThreadPool.AddGroupTask(Callable.From&lt;int&gt;(ProcessEnemyAI), enemies.Length);
		    // Other code...
		    WorkerThreadPool.WaitForTaskCompletion(taskID);
		    // Other code that depends on the enemy AI already being processed.
```

raulsntos reviewed (changes requested):
The descriptions seem more complete in this PR, so I would prefer this one. The code sample is simpler but I think that's better.

I made a few comments about invalid C# syntax, I can't really review the documentation itself since I'm not familiar with the API.

MewPurPur commented:
TY, I don't know C# but I tried my best. You should see the first version, I used `#` for comments XD

MewPurPur commented:
Fixed the issues, also you missed that in C#, I used the Task wait function, not the GroupTask version. I noticed it and fixed that too.

Changed the comment to just say that the array is supposed to be filled with enemies.

raulsntos commented on doc/classes/WorkerThreadPool.xml:
```diff
@@ -1,10 +1,50 @@
+		    # Other code that depends on the enemy AI already being processed.
+		[/gdscript]
+		[csharp]
+		private Node[] _enemies = {}; // An array to be filled with enemies.
```
Since C# arrays are immutable, I think it'd be better to use `List` here. Then, in `_Process` replace `enemies.Length` with `enemies.Count`.

It probably doesn't matter too much, but I think using a mutable collection would be better for a code example because then the user can change its size which the comment seems to suggest. Also, C# lists are more similar to GDScript's arrays than C# arrays. Sorry for making you change this again :disappointed:.

```suggestion
		private List&lt;Node&gt; _enemies = new List&lt;Node&gt;(); // An array to be filled with enemies.
```

MewPurPur replied:
So "A list to be filled with enemies" then?

raulsntos commented on doc/classes/WorkerThreadPool.xml:
```diff
@@ -1,10 +1,50 @@
+
+		public override void _Process(double delta)
+		{
+		    long taskID = WorkerThreadPool.AddGroupTask(Callable.From&lt;int&gt;(ProcessEnemyAI), enemies.Length);
```
Since `_enemies` is now a list, it should be using the `Count` property. Also noticed the missing underscore (`_`).

Also renamed `taskId` to follow .NET naming convention (ID is one of the [special cases](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#capitalizing-compound-words-and-common-terms)). Sorry for not noticing earlier.

```suggestion
		    long taskId = WorkerThreadPool.AddGroupTask(Callable.From&lt;int&gt;(ProcessEnemyAI), _enemies.Count);
```

RandomShaper commented:
Looks good! However, let me suggest documenting the various related project settings as well. There's a kind of obscure setting about using native threads for low-prio tasks (that I'm proposing getting rid of, by the way). In general, documenting the priority mechanics would be ideal.

raulsntos reviewed (approved):
Sorry this completely slipped my mind, the C# example is now correct.

akien-mga commented:
Thanks!

MewPurPur commented:
Lol I was just about to add those descriptions and I was confused where the PR went

YuriSizov commented:
Cherry-picked for 4.0.3.
